### PR TITLE
fix cv function metric bug

### PR DESCRIPTION
**Pull request by fstonezst**
#470

### CODE REVIEW

File: python-package/lightgbm/engine.py [hunk 1]
```diff
@@ -370,6 +370,8 @@ def cv(params, train_set, num_boost_round=10,
 
     if metrics:
         params.setdefault('metric', [])
+        if isinstance(params['metric'], string_type):
```
Comment (wxchan) [inline python-package/lightgbm/engine.py]: I suggest just override params['metric'] with metrics, params['metric'] actually can be any iterator

Comment (fstonezst) [inline python-package/lightgbm/engine.py]: If just override params['metric'] with metrics,then we will lose the metrics set in the params.
If metrics is a set , params['metric'].extend(metrics)  can also work.

Comment (wxchan) [inline python-package/lightgbm/engine.py]: I don't think user will set both params['metric'] and metrics intensionally, and expect both of them work. I guess most of time, user copied the params he used before and didn't notice 'metric' is in it, or can't figure out which to use, then set them with same thing and get duplicate outputs. So I think override and specify it in docs is enough.

Comment (fstonezst) [inline python-package/lightgbm/engine.py]: User set metrics in cv for evaluate each iterator is better or worse.
set metrics in param  for loss function.

Comment (wxchan) [inline python-package/lightgbm/engine.py]: metrics have nothing to do with loss function

Comment (fstonezst) [inline python-package/lightgbm/engine.py]: So how can we set different loss function for a specific task ? for example binary classification.

Comment (wxchan) [inline python-package/lightgbm/engine.py]: https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py#L121-L129

Comment (fstonezst) [inline python-package/lightgbm/engine.py]: OK，thanks! I get it. If you want to change the default loss function no matter what it is  you must write your own code. Then I agree with you to override and specify it in docs .  I will resubmit pull request you can choose one of them or fix it by yourself.

Comment (fstonezst) [inline python-package/lightgbm/engine.py]: I  closed the last pr and will add test case in test_cv(),thanks.

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -163,6 +163,9 @@ def test_cv(self):
         tss = TimeSeriesSplit(3)
         lgb.cv({'verbose': -1}, lgb_train, num_boost_round=10, data_splitter=tss, nfold=5,  # test if wrong nfold is ignored
                metrics='l2', verbose_eval=False)
+        # params has a string type metric and metrics is not None
+        lgb.cv({'verbose': -1, 'metric': 'binary_logloss'}, lgb_train, num_boost_round=10, nfold=3,
+               shuffle=False, metrics='auc', verbose_eval=False)
```
Comment (wxchan) [inline tests/python_package_test/test_engine.py]: the default task is regression, so metrics better be l1 or l2. I think you can re-use L156-157

**Comment by wxchan** [comment]
docstring should also be updated: https://github.com/fstonezst/LightGBM/blob/9822d2bf773d071059ee3819b0da547514810950/python-package/lightgbm/engine.py#L315

File: tests/python_package_test/test_engine.py (hunk 2 above)
Comment (fstonezst) [inline tests/python_package_test/test_engine.py]: OK  I've changed.

**Comment by fstonezst** [comment]
OMG! The Travis CI buil still failed  @wxchan 

log:
running test_engine.py
........E.
======================================================================
ERROR: test_regreesion (__main__.TestEngine)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_engine.py", line 75, in test_regreesion
    self.assertAlmostEqual(min(evals_result['eval']['l2']), ret, places=5)
ValueError: min() arg is an empty sequence
----------------------------------------------------------------------
Ran 10 tests in 1.922s
FAILED (errors=1)
The command "cd $TRAVIS_BUILD_DIR/tests/python_package_test && python test_basic.py && python test_engine.py && python test_sklearn.py && python test_plotting.py" exited with 1.

**Comment by wxchan** [comment]
you can revert the changes to test_engine.py, I will re-write some tests later.

btw, I suggest not to refer any issue in your commit message, it looks a little messy

**Comment by fstonezst** [comment]
OK , I see.

**Comment by fstonezst** [comment]
TAT，It's my honor to become a Contributor . Thank you for helping me so well. I'll have a sleepless night. 
No matter when your two guys come to Hangzhou, please send me an e-mail. Go for a drink, I'll treat.  TAT

**Pull request merged**